core/src/main/java/org/elasticsearch/indices/cluster/IndicesClusterStateService.java

```diff
@@ -267,376 +302,264 @@ protected void doRun() throws Exception {
-                    sendFailShard(shardRouting, "master [" + nodes.getMasterNode() + "] marked shard as started, but shard has not been created, mark shard as failed", null);
-                }
+        for (final ShardRouting shardRouting : localRoutingNode) {
+            ShardId shardId = shardRouting.shardId();
+            if (failedShardsCache.containsKey(shardId)) {
+                // don't create or update failed shards
                 continue;
             }
 
-            IndexShard indexShard = indexService.getShardOrNull(shardId);
-            if (indexShard != null) {
-                ShardRouting currentRoutingEntry = indexShard.routingEntry();
-                // if the current and global routing are initializing, but are still not the same, its a different "shard" being allocated
-                // for example: a shard that recovers from one node and now needs to recover to another node,
-                //              or a replica allocated and then allocating a primary because the primary failed on another node
-                boolean shardHasBeenRemoved = false;
-                assert currentRoutingEntry.isSameAllocation(shardRouting) :
-                    "local shard has a different allocation id but wasn't cleaning by applyDeletedShards. "
-                        + "cluster state: " + shardRouting + " local: " + currentRoutingEntry;
-                if (shardRouting.isPeerRecovery()) {
-                    RecoveryState recoveryState = indexShard.recoveryState();
-                    final DiscoveryNode sourceNode = findSourceNodeForPeerRecovery(logger, routingTable, nodes, shardRouting);
-                    if (recoveryState.getSourceNode().equals(sourceNode) == false) {
-                        if (recoveryTargetService.cancelRecoveriesForShard(currentRoutingEntry.shardId(), "recovery source node changed")) {
-                            // getting here means that the shard was still recovering
-                            logger.debug("[{}][{}] removing shard (recovery source changed), current [{}], global [{}])", shardRouting.index(), shardRouting.id(), currentRoutingEntry, shardRouting);
-                            indexService.removeShard(shardRouting.id(), "removing shard (recovery source node changed)");
-                            shardHasBeenRemoved = true;
-                        }
-                    }
-                }
-
-                if (shardHasBeenRemoved == false) {
-                    try {
-                        indexShard.updateRoutingEntry(shardRouting, event.state().blocks().disableStatePersistence() == false);
-                    } catch (Throwable e) {
-                        failAndRemoveShard(shardRouting, indexService, true, "failed updating shard routing entry", e);
-                    }
+            AllocatedIndex<? extends Shard> indexService = indicesService.indexService(shardId.getIndex());
+            assert indexService != null;
```

Comment:
can we remove this and just call `indexService.getShardOrNull` this will cause an NPE and we are good? I don't see why we need to protect here unless we throw IllegalState if we are null?